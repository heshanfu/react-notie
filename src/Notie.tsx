import * as React from 'react';
import * as ReactDOM from 'react-dom';

import {
    NotieActions,
    NotieSettings,
    NotieLevel,
    alert,
    NotieCallback,
    NotieEventDetails
} from './actions';

import {
    NotieContainer,
    NotieMessage,
    NotieChoices,
    NotieButton,
    NotieOverlay
} from './components';

const DEFAULT_STATE = {
    visible: false,
    locked: false,
    position: 'top',
    okBtnText: 'OK',
    cancelBtnText: 'Cancel'
};

export interface NotieProps {}

export interface NotieState extends Pick<NotieSettings, 'message' | 'level'> {
    visible: boolean;
    locked: boolean;
    okBtnText: string;
    cancelBtnText: string;
    position: 'top' | 'bottom';
}

let alreadyLoaded = false;

const DEFAULT_TIMEOUT = 5000;

export class Notie extends React.Component<NotieProps, NotieState> {
    private root: HTMLDivElement;
    private notieRoot: HTMLDivElement | null;
    private transitionendCallback: null | NotieCallback;
    private timeout: number | null;
    private confirmResolve: NotieCallback;
    private confirmReject: NotieCallback;

    constructor(props: NotieProps) {
        super(props);

        this.state = {
            message: '',
            level: NotieLevel.INFO,
            visible: false,
            locked: false,
            position: 'top',
            okBtnText: 'OK',
            cancelBtnText: 'Cancel'
        };
        this.root = document.createElement('div');
    }

    componentWillMount() {
        if (alreadyLoaded)
            throw new Error(
                '<Notie /> is already loaded somwhere in your code. Please include it only once.'
            );

        alreadyLoaded = true;
    }

    componentDidMount() {
        if (this.notieRoot) this.notieRoot.style.opacity = '0';
        window.addEventListener(NotieActions.SHOW, this.showNotie);
        document.body.appendChild(this.root);
    }

    shouldComponentUpdate(_: NotieProps, nextState: NotieState) {
        return this.state.visible !== nextState.visible;
    }

    componentWillUnmount() {
        if (this.transitionendCallback) {
            this.notieRoot &&
                this.notieRoot.removeEventListener('transitionend', this.transitionendCallback);
            this.transitionendCallback = null;
        }

        if (this.timeout) {
            clearTimeout(this.timeout);
            this.timeout = null;
        }

        if (alreadyLoaded) alreadyLoaded = false;

        window.removeEventListener(NotieActions.SHOW, this.showNotie);
        document.body.removeChild(this.root);
    }

    showNotie = (event: Event) => {
        if (this.state.locked) return;

        const settings: NotieEventDetails = (event as CustomEvent).detail;
        const toLock = settings.level === NotieLevel.CONFIRM || settings.level === NotieLevel.FORCE;

        if (settings.confirmReject && settings.confirmResolve) {
            this.confirmReject = settings.confirmReject;
            this.confirmResolve = settings.confirmResolve;
        } else {
            delete this.confirmReject;
            delete this.confirmResolve;
        }

        if (this.timeout) {
            clearTimeout(this.timeout);
            this.timeout = null;
        }

        if (this.state.visible) {
            this.hideNotie(() => alert(settings));
            return;
        }

        if (this.notieRoot) this.notieRoot.style.opacity = '1';

        this.setState(state => ({
            message: settings.message,
            level: settings.level,
            visible: true,
            locked: toLock,
            okBtnText: settings.yesBtnText || state.okBtnText,
            cancelBtnText: settings.noBtnText || state.cancelBtnText
        }));

        if (!toLock) {
            this.timeout = window.setTimeout(() => {
                this.hideNotie();
                this.timeout = null;
            }, settings.ttl || DEFAULT_TIMEOUT);
        }
    };

    hideNotie = (callback?: NotieCallback) => {
        this.setState(DEFAULT_STATE as NotieState);
        this.transitionendCallback = () => {
            if (this.notieRoot) this.notieRoot.style.opacity = '0';
            if (typeof callback === 'function') callback();
            this.notieRoot &&
                this.notieRoot.removeEventListener('transitionend', this
                    .transitionendCallback as NotieCallback);
            this.transitionendCallback = null;
        };
        this.notieRoot &&
            this.notieRoot.addEventListener('transitionend', this.transitionendCallback);
    };

    handleYes = () => this.hideNotie(this.confirmResolve);

    handleNo = () => this.hideNotie(this.confirmReject);

    handleDismiss = () => {
        if (this.state.locked) return;

        if (this.timeout) {
            clearTimeout(this.timeout);
            this.timeout = null;
        }

        if (this.state.visible) {
            this.hideNotie();
        }
    };

    rootRef = (c: HTMLDivElement) => (this.notieRoot = c);

    render() {
        const { message, level, visible, position } = this.state;
        const forced = level === NotieLevel.CONFIRM || level === NotieLevel.FORCE;

        return ReactDOM.createPortal(
            <div ref={this.rootRef}>
                {forced && visible && <NotieOverlay />}
                <NotieContainer
                    active={visible}
                    position={position}
                    level={level}
                    onClick={this.handleDismiss}>
                    <NotieMessage>{message}</NotieMessage>
                    {forced && (
                        <NotieChoices>
                            <NotieButton
                                level={
                                    level === NotieLevel.FORCE
                                        ? NotieLevel.ERROR
                                        : NotieLevel.SUCCESS
                                }
                                onClick={this.handleYes}>
                                {this.state.okBtnText}
                            </NotieButton>
                            {level === NotieLevel.CONFIRM && (
                                <NotieButton level={NotieLevel.ERROR} onClick={this.handleNo}>
                                    {this.state.cancelBtnText}
                                </NotieButton>
                            )}
                        </NotieChoices>
                    )}
                </NotieContainer>
            </div>,
            this.root
        );
    }
}
