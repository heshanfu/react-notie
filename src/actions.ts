export enum NotieLevel {
    INFO = 'INFO',
    ERROR = 'ERROR',
    SUCCESS = 'SUCCESS',
    WARN = 'WARN',
    FORCE = 'FORCE',
    CONFIRM = 'CONFIRM'
}

export enum NotieActions {
    SHOW = 'notie.show',
    HIDE = 'notie.hide'
}

export interface NotieSettings {
    message: string;
    level: NotieLevel;
    yesBtnText?: string;
    noBtnText?: string;
    ttl?: number;
}

export type NotieCallback = (...args: any[]) => void;

export interface NotieEventDetails extends NotieSettings {
    confirmResolve: NotieCallback;
    confirmReject: NotieCallback;
}

export function dispatchGlobalEvent(eventName: string, opts: any, target: EventTarget = window) {
    // Compatibale with IE
    // @see http://stackoverflow.com/questions/26596123/internet-explorer-9-10-11-event-constructor-doesnt-work
    let event;

    /* istanbul ignore else */
    if (typeof (<any>window).CustomEvent === 'function') {
        event = new (<any>window).CustomEvent(eventName, { detail: opts });
    } else {
        event = document.createEvent('CustomEvent');
        event.initCustomEvent(eventName, false, true, opts);
    }

    /* istanbul ignore else */
    if (target) {
        target.dispatchEvent(event);
    }
}

export function alert(settings: NotieSettings) {
    dispatchGlobalEvent(NotieActions.SHOW, settings);
}

export function success(message: string, ttl?: number) {
    alert({ message, level: NotieLevel.SUCCESS, ttl });
}

export function error(message: string, ttl?: number) {
    alert({ message, level: NotieLevel.ERROR, ttl });
}

export function warn(message: string, ttl?: number) {
    alert({ message, level: NotieLevel.WARN, ttl });
}

export function info(message: string, ttl?: number) {
    alert({ message, level: NotieLevel.INFO, ttl });
}

export function confirm(message: string, props?: Partial<NotieSettings>) {
    return new Promise((resolve, reject) => {
        alert(
            Object.assign({}, props, {
                message,
                level: NotieLevel.CONFIRM,
                confirmResolve: resolve,
                confirmReject: reject
            })
        );
    });
}

export function force(message: string, props?: Partial<NotieSettings>) {
    return new Promise((resolve, reject) => {
        alert(
            Object.assign({}, props, {
                message,
                level: NotieLevel.FORCE,
                confirmResolve: resolve,
                confirmReject: reject
            })
        );
    });
}
