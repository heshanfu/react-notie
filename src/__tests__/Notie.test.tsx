import * as React from 'react';
import { mount } from 'enzyme';

import { Notie } from '../Notie';
import * as actions from '../actions';

jest.useFakeTimers();

describe('<Notie />', () => {
    let component;

    beforeEach(() => {
        component = mount(<Notie />);
    });

    afterEach(() => {
        if (component) {
            component.unmount();
            component = null;
        }
    });

    test('initial render', () => {
        expect(component).toMatchSnapshot();
    });

    test('can be mounted only once', () => {
        expect(() => mount(<Notie />)).toThrow();
    });

    test('success message', () => {
        actions.success('Success!');
        component.update();
        expect(component.find('.notie--success').length).toBe(1);
        expect(component.find('.notie--active').length).toBe(1);
        expect(component).toMatchSnapshot();
        jest.runTimersToTime(5000);
        component.update();
        expect(component.find('.notie--active').length).toBe(0);
        expect(component).toMatchSnapshot();
    });

    test('warn message', () => {
        actions.warn('Warn!');
        component.update();
        expect(component.find('.notie--warn').length).toBe(1);
        expect(component.find('.notie--active').length).toBe(1);
        expect(component).toMatchSnapshot();
        jest.runTimersToTime(5000);
        component.update();
        expect(component.find('.notie--active').length).toBe(0);
        expect(component).toMatchSnapshot();
    });

    test('error message', () => {
        actions.error('Error!');
        component.update();
        expect(component.find('.notie--error').length).toBe(1);
        expect(component.find('.notie--active').length).toBe(1);
        expect(component).toMatchSnapshot();
        jest.runTimersToTime(5000);
        component.update();
        expect(component.find('.notie--active').length).toBe(0);
        expect(component).toMatchSnapshot();
    });

    test('info message', () => {
        actions.info('Info!');
        component.update();
        expect(component.find('.notie--info').length).toBe(1);
        expect(component.find('.notie--active').length).toBe(1);
        expect(component).toMatchSnapshot();
        jest.runTimersToTime(5000);
        component.update();
        expect(component.find('.notie--active').length).toBe(0);
        expect(component).toMatchSnapshot();
    });

    test('consecutive calls', () => {
        const instance = component.instance();
        actions.info('Info!');
        component.update();
        expect(component.find('.notie--info').length).toBe(1);
        expect(component.find('.notie--active').length).toBe(1);
        expect(component).toMatchSnapshot();
        jest.runTimersToTime(2000);
        component.update();
        actions.error('Error!');

        /*
            Must be triggered manually since `transitionend` event
            is not being trigged when using `dispatchEvent` on root
            element.
        */
        instance.transitionendCallback();
        expect(instance.transitionendCallback).toBe(null);
        component.update();

        expect(component.find('.notie--error').length).toBe(1);
        expect(component.find('.notie--active').length).toBe(1);
        expect(component).toMatchSnapshot();
    });

    test('cleanup on unmount', () => {
        const instance = component.instance();
        actions.info('Info!');
        actions.info('Info!');
        jest.runTimersToTime(2000);
        component.update();
        component.unmount();
        expect(instance.transitionendCallback).toBe(null);
    });

    test('dismiss on click', () => {
        actions.info('Info!');
        component.update();
        expect(component.find('.notie--info').length).toBe(1);
        expect(component.find('.notie--active').length).toBe(1);
        expect(component).toMatchSnapshot();
        component.find('.notie-container').simulate('click');
        component.update();
        expect(component.find('.notie--active').length).toBe(0);
        expect(component).toMatchSnapshot();
    });

    describe('confirm', () => {
        test('initial render', () => {
            actions.confirm('Are you sure?');
            component.update();
            expect(component.find('.notie--confirm').length).toBe(1);
            expect(component.find('.notie--active').length).toBe(1);
            expect(component.find('.notie-btn--ok').length).toBe(1);
            expect(component.find('.notie-btn--cancel').length).toBe(1);
            expect(component).toMatchSnapshot();
        });

        test('Resolves on postive feedback', done => {
            const instance = component.instance();
            const promise = actions.confirm('Are you sure?');

            component.update();
            component.find('.notie-btn--ok').simulate('click');
            instance.transitionendCallback();
            expect(instance.transitionendCallback).toBe(null);
            component.update();

            promise.then(() => {
                expect(component).toMatchSnapshot();
                expect(component.find('.notie--active').length).toBe(0);
                done();
            });
        });

        test('Rejects on negative feedback', done => {
            const instance = component.instance();
            const promise = actions.confirm('Are you sure?');

            component.update();
            component.find('.notie-btn--cancel').simulate('click');
            instance.transitionendCallback();
            expect(instance.transitionendCallback).toBe(null);
            component.update();
            promise.catch(() => {
                expect(component).toMatchSnapshot();
                expect(component.find('.notie--active').length).toBe(0);
                done();
            });
        });

        test('Consecutive calls do not work', () => {
            const instance = component.instance();
            actions.confirm('Are you sure?');
            component.update();
            expect(component.find('.notie--confirm').length).toBe(1);
            expect(component).toMatchSnapshot();
            actions.info('Info!');
            component.update();
            expect(component.find('.notie--confirm').length).toBe(1);
            expect(component).toMatchSnapshot();
        });

        test('Custom button text', () => {
            const instance = component.instance();
            const options = {
                yesBtnText: 'Hell Yeah!',
                noBtnText: 'Fuck NO!'
            };
            actions.confirm('Are you sure?', options);
            component.update();
            expect(component.find('.notie-btn--ok').text()).toBe(options.yesBtnText);
            expect(component.find('.notie-btn--cancel').text()).toBe(options.noBtnText);
            expect(component).toMatchSnapshot();
        });
    });

    describe('force', () => {
        test('initial render', () => {
            actions.force('You have done it again!');
            component.update();
            expect(component.find('.notie--force').length).toBe(1);
            expect(component.find('.notie--active').length).toBe(1);
            expect(component.find('.notie-btn--ok').length).toBe(1);
            expect(component.find('.notie-btn--cancel').length).toBe(0);

            expect(component).toMatchSnapshot();
        });

        test('Resolves on feedback', done => {
            const instance = component.instance();
            const promise = actions.force('You have done it again!');

            component.update();
            component.find('.notie-btn--ok').simulate('click');
            instance.transitionendCallback();
            expect(instance.transitionendCallback).toBe(null);
            component.update();

            promise.then(() => {
                expect(component).toMatchSnapshot();
                expect(component.find('.notie--active').length).toBe(0);
                done();
            });
        });

        test('Consecutive calls do not work', () => {
            const instance = component.instance();
            actions.force('You have done it again!');
            component.update();
            expect(component.find('.notie--force').length).toBe(1);
            expect(component).toMatchSnapshot();
            actions.info('Info!');
            component.update();
            expect(component.find('.notie--force').length).toBe(1);
            expect(component).toMatchSnapshot();
        });

        test('Custom button text', () => {
            const instance = component.instance();
            const options = {
                yesBtnText: 'Hmmm OK?'
            };
            actions.confirm('Are you sure?', options);
            component.update();
            expect(component.find('.notie-btn--ok').text()).toBe(options.yesBtnText);
            expect(component).toMatchSnapshot();
        });
    });
});
