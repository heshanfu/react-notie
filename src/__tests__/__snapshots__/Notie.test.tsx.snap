// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Notie /> Error message 1`] = `
.emotion-0 {
  padding: 20px;
}

.emotion-1 {
  position: fixed;
  width: 100%;
  text-align: center;
  font-size: 14px;
  color: #fff;
  margin: 0 auto;
  white-space: nowrap;
  border: none;
  text-overflow: ellipsis;
  left: 0;
  top: 0;
  padding: 0;
  pointer-events: none;
  background-color: #ff4949;
}

.emotion-1[open] {
  pointer-events: auto;
}

<Notie>
  <NotieContainer
    innerRef={[Function]}
    level="ERROR"
    onClick={[Function]}
  >
    <dialog
      className="emotion-1"
      onClick={[Function]}
    >
      <NotieMessage>
        <div
          className="emotion-0"
        >
          Error!
        </div>
      </NotieMessage>
    </dialog>
  </NotieContainer>
</Notie>
`;

exports[`<Notie /> Error message 2`] = `
.emotion-0 {
  padding: 20px;
}

.emotion-1 {
  position: fixed;
  width: 100%;
  text-align: center;
  font-size: 14px;
  color: #fff;
  margin: 0 auto;
  white-space: nowrap;
  border: none;
  text-overflow: ellipsis;
  left: 0;
  top: 0;
  padding: 0;
  pointer-events: none;
  background-color: #ff4949;
}

.emotion-1[open] {
  pointer-events: auto;
}

<Notie>
  <NotieContainer
    innerRef={[Function]}
    level="ERROR"
    onClick={[Function]}
  >
    <dialog
      className="emotion-1"
      onClick={[Function]}
    >
      <NotieMessage>
        <div
          className="emotion-0"
        >
          Error!
        </div>
      </NotieMessage>
    </dialog>
  </NotieContainer>
</Notie>
`;

exports[`<Notie /> info message 1`] = `
.emotion-1 {
  position: fixed;
  width: 100%;
  text-align: center;
  font-size: 14px;
  color: #fff;
  margin: 0 auto;
  white-space: nowrap;
  border: none;
  text-overflow: ellipsis;
  left: 0;
  top: 0;
  padding: 0;
  pointer-events: none;
  background-color: #20a0ff;
}

.emotion-1[open] {
  pointer-events: auto;
}

.emotion-0 {
  padding: 20px;
}

<Notie>
  <NotieContainer
    innerRef={[Function]}
    level="INFO"
    onClick={[Function]}
  >
    <dialog
      className="emotion-1"
      onClick={[Function]}
    >
      <NotieMessage>
        <div
          className="emotion-0"
        >
          Info!
        </div>
      </NotieMessage>
    </dialog>
  </NotieContainer>
</Notie>
`;

exports[`<Notie /> info message 2`] = `
.emotion-1 {
  position: fixed;
  width: 100%;
  text-align: center;
  font-size: 14px;
  color: #fff;
  margin: 0 auto;
  white-space: nowrap;
  border: none;
  text-overflow: ellipsis;
  left: 0;
  top: 0;
  padding: 0;
  pointer-events: none;
  background-color: #20a0ff;
}

.emotion-1[open] {
  pointer-events: auto;
}

.emotion-0 {
  padding: 20px;
}

<Notie>
  <NotieContainer
    innerRef={[Function]}
    level="INFO"
    onClick={[Function]}
  >
    <dialog
      className="emotion-1"
      onClick={[Function]}
    >
      <NotieMessage>
        <div
          className="emotion-0"
        >
          Info!
        </div>
      </NotieMessage>
    </dialog>
  </NotieContainer>
</Notie>
`;

exports[`<Notie /> initial render 1`] = `
.emotion-1 {
  position: fixed;
  width: 100%;
  text-align: center;
  font-size: 14px;
  color: #fff;
  margin: 0 auto;
  white-space: nowrap;
  border: none;
  text-overflow: ellipsis;
  left: 0;
  top: 0;
  padding: 0;
  pointer-events: none;
  background-color: #20a0ff;
}

.emotion-1[open] {
  pointer-events: auto;
}

.emotion-0 {
  padding: 20px;
}

<Notie>
  <NotieContainer
    innerRef={[Function]}
    level="INFO"
    onClick={[Function]}
  >
    <dialog
      className="emotion-1"
      onClick={[Function]}
    >
      <NotieMessage>
        <div
          className="emotion-0"
        />
      </NotieMessage>
    </dialog>
  </NotieContainer>
</Notie>
`;

exports[`<Notie /> success message 1`] = `
.emotion-0 {
  padding: 20px;
}

.emotion-1 {
  position: fixed;
  width: 100%;
  text-align: center;
  font-size: 14px;
  color: #fff;
  margin: 0 auto;
  white-space: nowrap;
  border: none;
  text-overflow: ellipsis;
  left: 0;
  top: 0;
  padding: 0;
  pointer-events: none;
  background-color: #13ce66;
}

.emotion-1[open] {
  pointer-events: auto;
}

<Notie>
  <NotieContainer
    innerRef={[Function]}
    level="SUCCESS"
    onClick={[Function]}
  >
    <dialog
      className="emotion-1"
      onClick={[Function]}
    >
      <NotieMessage>
        <div
          className="emotion-0"
        >
          Success!
        </div>
      </NotieMessage>
    </dialog>
  </NotieContainer>
</Notie>
`;

exports[`<Notie /> success message 2`] = `
.emotion-0 {
  padding: 20px;
}

.emotion-1 {
  position: fixed;
  width: 100%;
  text-align: center;
  font-size: 14px;
  color: #fff;
  margin: 0 auto;
  white-space: nowrap;
  border: none;
  text-overflow: ellipsis;
  left: 0;
  top: 0;
  padding: 0;
  pointer-events: none;
  background-color: #13ce66;
}

.emotion-1[open] {
  pointer-events: auto;
}

<Notie>
  <NotieContainer
    innerRef={[Function]}
    level="SUCCESS"
    onClick={[Function]}
  >
    <dialog
      className="emotion-1"
      onClick={[Function]}
    >
      <NotieMessage>
        <div
          className="emotion-0"
        >
          Success!
        </div>
      </NotieMessage>
    </dialog>
  </NotieContainer>
</Notie>
`;

exports[`<Notie /> warn message 1`] = `
.emotion-0 {
  padding: 20px;
}

.emotion-1 {
  position: fixed;
  width: 100%;
  text-align: center;
  font-size: 14px;
  color: #fff;
  margin: 0 auto;
  white-space: nowrap;
  border: none;
  text-overflow: ellipsis;
  left: 0;
  top: 0;
  padding: 0;
  pointer-events: none;
  background-color: #f5be4b;
}

.emotion-1[open] {
  pointer-events: auto;
}

<Notie>
  <NotieContainer
    innerRef={[Function]}
    level="WARN"
    onClick={[Function]}
  >
    <dialog
      className="emotion-1"
      onClick={[Function]}
    >
      <NotieMessage>
        <div
          className="emotion-0"
        >
          Warn!
        </div>
      </NotieMessage>
    </dialog>
  </NotieContainer>
</Notie>
`;

exports[`<Notie /> warn message 2`] = `
.emotion-0 {
  padding: 20px;
}

.emotion-1 {
  position: fixed;
  width: 100%;
  text-align: center;
  font-size: 14px;
  color: #fff;
  margin: 0 auto;
  white-space: nowrap;
  border: none;
  text-overflow: ellipsis;
  left: 0;
  top: 0;
  padding: 0;
  pointer-events: none;
  background-color: #f5be4b;
}

.emotion-1[open] {
  pointer-events: auto;
}

<Notie>
  <NotieContainer
    innerRef={[Function]}
    level="WARN"
    onClick={[Function]}
  >
    <dialog
      className="emotion-1"
      onClick={[Function]}
    >
      <NotieMessage>
        <div
          className="emotion-0"
        >
          Warn!
        </div>
      </NotieMessage>
    </dialog>
  </NotieContainer>
</Notie>
`;
